Maintain TCP connections/sessions independently

myleshorton opened the pull request:
This is for https://github.com/getlantern/lantern-internal/issues/2967 and works in conjunction with https://github.com/getlantern/flashlight/pull/664 to maintain TCP connections/lampshade sessions automatically in the background.

myleshorton commented:
I'm still playing around with this @joesis -- should have something tomorrow!

myleshorton commented:
OK what do you think about this @joesis? We basically maintain two TCP connections to each lampshade server, one with a short dial timeout and one with a longer one. The shorter one should ensure liveness in the case of network disruptions, and its dial time is big enough in practice for to be a backup connection in most situations.

joesis commented on dialer.go:
```diff
@@ -81,137 +84,157 @@ func NewDialer(opts *DialerOpts) Dialer {
+		emaRTT:                ema.NewDuration(0, 0.5),
+		dial:                  opts.Dial,
+		pendingSessions:       make(chan *pendingSession, opts.LiveConns),
+		lifecyle:              opts.Lifecycle,
```
Oh this can be nil which called by proxybench which causes panic in L108.

myleshorton commented:
Thanks @joesis will address!

myleshorton commented on session.go:
```diff
@@ -582,42 +577,21 @@ func (s *session) getOrCreateStream(id uint16) (*stream, bool) {
 	s.closed[id] = true
-	if s.defunct && len(s.streams) == 0 {
-		s.Close()
-	}
```
So I tweaked AllowNewStreams to only check if we've reached the maximum number of streams, which will almost never happen in practice. If it does, we'll create a new session, but we don't worry about all the 65K streams on a given session closing and then closing the session because both ever reaching 65K and all 65K closing in a given lantern run are super unlikely, and if it does we just have a dangling TCP connection that will get closed by the server on idle timeout.

In the other case that the current session is closed, we obviously don't have to worry about closing it again.

myleshorton commented on lifecycle.go:
```diff
@@ -0,0 +1,105 @@
+
+// LifecycleListener allows lampshade users to listen to lampshade lifecycle events.
+type LifecycleListener interface {
+	OnSessionInit(context.Context) context.Context
```
Yeah that was actually my original thought especially with `StreamLifecycleListener` -- let me play with that again. I got away from it because the holding state seemed to cause issues with traces and thread safety, but I think that's only the case if we store spans, not contexts.

myleshorton commented on stream.go:
```diff
@@ -22,16 +23,18 @@ type stream struct {
+		pool:      bp,
+		sb:        newSendBuffer(defaultHeader, w, windowSize),
+		rb:        newReceiveBuffer(defaultHeader, w, bp, windowSize),
+		lifecycle: s.lifecycle.OnStreamInit(s.ctx, ctx, id),
```
yeah just changed to use `context.Background` on `getOrCreateStream` calls from the read loop, which IMU are only called from the server side.

myleshorton replied:
Currently this implementation does not support tracing at the lampshade protocol level on the server -- my current thought is to purely trace HTTP on that side.

myleshorton commented:
OK @joesis I tweaked the listener interfaces considerably to avoid passing contexts around quite so much, and it seems to work. I also fixed your other suggestions I believe.

What do you think?

joesis commented on dialer.go:
```diff
@@ -81,137 +83,161 @@ func NewDialer(opts *DialerOpts) Dialer {
+					log.Debug("Added required session")
+				default:
+					log.Debug("Maximum sessions reached")
+				}
```
Since `s. allowNewStream` should rarely be `false`, it may be worth printing one log message explicitly, or, change the above debug prints to emphasize that it's the result of max streams per conn being hit.

joesis replied:
Shouldn't it `continue` to get the next live session? The current session can't create new streams anyway.

myleshorton replied:
Yes great catch on the continue!

joesis commented on session.go:
```diff
@@ -151,7 +150,15 @@ func (s *session) recvLoop() {
+			s.requiredSessions <- s.requiredSession
+		}
 		closeErr := s.Conn.Close()
+		if s.lifecycle != nil {
```
Not a bit deal, but as we already have `NoopClientLifecycleListener`, maybe pass it when start session in `listener.go` too, so this check can be removed.

joesis replied:
And potentially avoid panic in other places too

myleshorton replied:
good idea -- will do!

joesis commented:
Very cool! Just two more comments

myleshorton commented:
OK thanks @joesis! I just made those two fixes and added a callback in `onSessionError` as well.

joesis commented on dialer.go:
```diff
@@ -81,137 +83,161 @@ func NewDialer(opts *DialerOpts) Dialer {
+	d.pendingSessions <- &pendingSession{
+		name:         "liveness",
+		dialTimeout:  5 * time.Second,
+		sleepOnError: 0 * time.Second,
```
Oh I just realized when seeing logs - if dialing lampshade gets RST or errors like no route to host/network, Lantern would busily dial new session over and over again without sleeping.

joesis commented:
Here are the sequence of session requests. Somehow the `liveness` session is not created as expected when network is down (see logs from 33m to 35m).

```
 1m35s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}  <=== switching network
 1m35s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 1m40s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 1m40s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 1m45s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 1m46s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 2m55s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 3m0s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 3m12s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 4m42s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 6m29s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 9m33s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 11m4s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 12m34s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 20m41s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 22m11s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 23m41s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}

 33m17s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}  <=== disconnecting router from the Internet
 34m20s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}

 35m33s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}  <=== switching to a working network
 35m40s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 35m41s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 35m43s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 35m44s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
38m12s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 38m13s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 38m13s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 38m14s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 40m0s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 40m15s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 40m15s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
 40m31s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"background", dialTimeout:30000000000, sleepOnError:2000000000}
 40m32s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
```

joesis commented:
As we are here, it would be helpful if we can attach the name when creating lampshade dialer, to help correlate the logs to the proxy. See below.
```
=> which proxy is the new session to? When logs are interleaving, it's hard to know.
Sep 04 11:47:27.822 - 1m35s DEBUG lampshade: session.go:155 Requesting new session: &lampshade.pendingSession{name:"liveness", dialTimeout:5000000000, sleepOnError:0}
Sep 04 11:47:27.822 - 1m35s DEBUG chained: proxy.go:307 Dialing lampshade TCP connection to fp-anhk1bu16-20190828-037-lampshade
Sep 04 11:47:27.864 - 1m35s DEBUG chained: proxy.go:313 Successfully created lampshade TCP connection to fp-anhk1bu16-20190828-037-lampshade in 0.041763977 seconds
Sep 04 11:47:27.866 - 1m35s DEBUG lampshade: dialer.go:173 Created session in 44.377622ms
=> which session is closed?
Sep 04 11:47:27.871 - 1m35s DEBUG lampshade: session.go:153 Closing lampshade TCP connection
```

myleshorton commented:
Yup great catches again -- on it!

joesis commented:
Somehow the time between dialing each "liveness" session is longer than configured 5s.

```
108m35s DEBUG lampshade: dialer.go:169 Error starting session 'background': dial tcp 207.46.138.76:31301: i/o timeout
 108m42s DEBUG lampshade: dialer.go:169 Error starting session 'liveness': dial tcp 207.46.138.76:31301: i/o timeout
 109m12s DEBUG lampshade: dialer.go:169 Error starting session 'background': dial tcp 207.46.138.76:31301: i/o timeout
 109m19s DEBUG lampshade: dialer.go:169 Error starting session 'liveness': dial tcp 207.46.138.76:31301: i/o timeout
 109m49s DEBUG lampshade: dialer.go:169 Error starting session 'background': dial tcp 207.46.138.76:31301: i/o timeout
 109m56s DEBUG lampshade: dialer.go:169 Error starting session 'liveness': dial tcp 207.46.138.76:31301: i/o timeout
 110m26s DEBUG lampshade: dialer.go:169 Error starting session 'background': dial tcp 207.46.138.76:31301: i/o timeout
 110m33s DEBUG lampshade: dialer.go:169 Error starting session 'liveness': dial tcp 207.46.138.76:31301: i/o timeout
```

joesis commented:
Oh it's expected as there's only one `maintainTCPConnections` for each proxy. It needs to wait for (20s max) a "background" session to fail if there's one ongoing, sleep for 2s, then handle the next "liveness" request, rather than redialing at a regular 5s rate.

myleshorton commented:
> Oh it's expected as there's only one maintainTCPConnections for each proxy. It needs to wait for (20s max) a "background" session to fail if there's one ongoing, sleep for 2s, then handle the next "liveness" request, rather than redialing at a regular 5s rate.

Oh great catch again @joesis. Just seeing this now, but I'll address that in the next pass (likely tomorrow).

I just added the other changes. I also added an alternative byte buffer pool implementation backed with sync.Pool that's faster in the new benchmarks in lampshade_test.go.

Otherwise, I still need to fix the other tests.

myleshorton commented:
Actually just added a fix for that. What do you think?

joesis commented:
The new pool is a bit faster but involves one alloc per op.
```
BenchmarkSyncPool-4   	20000000	        70.9 ns/op	      32 B/op	       1 allocs/op
BenchmarkPool-4       	20000000	        87.3 ns/op	       0 B/op	       0 allocs/op
```

joesis commented:
Fully LGTM! Lampshade dialer recovers much more quickly now. I'm not sure how much performance gain `NewBufferSyncPool` can bring but @myleshorton feel free to swap it in or not.

joesis commented:
Wait I saw two "Maximum live sessions reached" message within one minute. There must be something wrong...

myleshorton commented:
OK I'm going to merge this in and build a new version. Also going to look at https://github.com/getlantern/lantern-internal/issues/2977

joesis commented on dialer.go:
```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
+		d.pendingSessions <- ps
+	} else {
+		log.Debugf("Created session in %v to %#v", time.Since(start), ps)
+		d.liveSessions <- s
```
Hmm right now when my Lantern resumes from disconnected network a lot of goroutines are blocked here. How did these requests sneak into `pendingSessions` in the first place?

```
goroutine profile: total 651
289 @ 0x40324ff 0x4007a53 0x4007a29 0x40077a5 0x4740db9 0x4061591
#	0x4740db8	github.com/getlantern/lampshade.(*dialer).trySession+0x418	/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190904225539-ee3e36398b21/dialer.go:148

76 @ 0x40324ff 0x404309b 0x4728c3f 0x4061591
#	0x4728c3e	github.com/getlantern/idletiming.Conn.func1+0x1fe	/Users/joesis/go/pkg/mod/github.com/getlantern/idletiming@v0.0.0-20190529182719-d2fbc83372a5/idletiming_conn.go:64
```

joesis replied:
```
goroutine profile: total 2003
1669 @ 0x40324ff 0x4007a53 0x4007a29 0x40077a5 0x4740db9 0x4061591
#	0x4740db8	github.com/getlantern/lampshade.(*dialer).trySession+0x418	/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190904225539-ee3e36398b21/dialer.go:148
```

joesis replied:
Okay here's what happened:
1.  When closing a lampshade session, it won't be removed from `d.liveSessions` until next `d.DialContext`, which means:
1. If the dialer is never selected again by balancer, `d.liveSessions` will never have room for the newly created session, so `trySession` blocks on sending to `d.liveSessions`.
1. The new session can also have error or be idle closed, which triggers yet another new session, and again, blocks on sending to `d.liveSessions`.
1. The same happens over and over again, causes tons of goroutines waiting to add already closed sessions to `d.liveSessions`.

A naive solution would be looping over `d.liveSessions` to remove any closed ones before adding the new one, but there are definitely better ways.

These rounds of bug hunting also made me wondering if the current logic is solid enough so we can just improve over it?

BTW `s.onSessionError` can be called for one session several times, hence `s.Close()`, that may be why requesting new session in `s.Close()` didn't work. If `s.Close()` is protected by `sync.Once`, the closing logic can be moved there.

myleshorton replied:
Ah great catch again @joesis! I still think this approach is significantly easier to work with in the end. I added a first pass at a solution but not quite ready for review yet!

myleshorton replied:
Gonna play with making it a state machine

myleshorton replied:
OK I just added a `recycleSession` method that should fix that particular issue @joesis. I'm super tempted to just turn the whole sequence into a state machine, but that's a bigger change. Might just leave it for now.

joesis commented on lampshade.go:
```diff
@@ -437,3 +409,31 @@ func echo() []byte {
 	return echo
 }
+
+type pendingSession struct {
```
Nit: Log entries like `Closing lampshade TCP connection: &lampshade.pendingSession{...` is a bit confusing. Maybe rename to `sessionConf` or so?

myleshorton replied:
OK

oxtoacart commented on dialer.go:
```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
-	if opts.MaxLiveConns <= 0 {
-		opts.MaxLiveConns = 1
+	if opts.LiveConns <= 0 {
+		opts.LiveConns = 2
```
The higher the number of live conns, the less likely we are to coalesce frames. Any reason not to default to 1?

myleshorton replied:
The second connection has a much shorter dial timeout to handle temporary network disconnections or jumps in congestion to ensure lantern recovers more quickly.

The inclusion of that second connection also just gives it some redundancy such that it's more likely one of the connections will be functioning especially at congested times.

So we could switch to a single connection, but we'd need to either revert to the previous code to ensure liveness or come up with a new approach. To me, though, the advantages outweigh the disadvantages.

oxtoacart replied:
> The inclusion of that second connection also just gives it some redundancy such that it's more likely one of the connections will be functioning especially at congested times.

If only one of the connections is actually functioning, what keeps us from picking the wrong connection for new streams?

myleshorton replied:
Sessions are only available when their TCP connections are active and the session is established. Streams certainly can and often do fail after that point, but the second connection at least gives dialers something to work with in the case the second connection has fully failed.

oxtoacart commented on dialer.go:
```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
 	}
+	lc.OnTCPEstablished(conn)
 
 	cs, err := newCryptoSpec(d.cipherCode)
```
This is not a new thing, but I noticed that if we get an error here or later, we don't close the conn. I think we should. It's likely not a problem in practice because we should never actually get errors here.

myleshorton replied:
nice catch will fix!

myleshorton replied:
I ended up just moving those calls up to before we create the TCP connection.

oxtoacart commented on dialer.go:
```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
+		dialTimeout:  5 * time.Second,
+		sleepOnError: 1 * time.Second,
 	}
+	go d.maintainTCPConnections()
```
Better to use `ops.Go` to maintain any ops context if needed.

myleshorton replied:
Cool will do

oxtoacart commented on dialer.go:
```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
+// maintainTCPConnections maintains background TCP connection(s) and associated lampshade session(s)
+func (d *dialer) maintainTCPConnections() {
+	for rs := range d.pendingSessions {
+		go d.trySession(rs)
```
Better to use `ops.Go` to maintain any ops context if needed.

myleshorton replied:
will do

oxtoacart commented on dialer.go:
```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
+
+	// We create another background session with a shorter dial timeout to ensure liveness in the case of network
+	// disruptions.
+	d.pendingSessions <- &pendingSession{
```
Why is this not using `newPendingSession`? Also, should we make dialTimeout and sleepOnError configurable?

myleshorton replied:
we'd have to add more arguments to `newPendingSession` but certainly could do that.

oxtoacart commented on session.go:
```diff
@@ -151,14 +150,20 @@ func (s *session) recvLoop() {
+		log.Debugf("Closing lampshade TCP connection: %#v", s.pendingSession)
+		if s.pendingSessions != nil {
+			log.Debugf("Requesting new session: %#v", s.pendingSession)
+			s.pendingSessions <- s.pendingSession
```
What if pendingSessions is full?

myleshorton replied:
`pendingSessions` is continually offloaded in `maintainTCPConnections`, so I don't think that can happen.

myleshorton commented:
OK, I believe all the comments/issues are now addressed.

myleshorton commented:
This is working super well for me FWIW.

joesis commented on dialer.go:
```diff
@@ -68,175 +81,168 @@ func NewDialer(opts *DialerOpts) Dialer {
+}
+
+func (d *dialer) getSession(ctx context.Context) (sessionIntf, error) {
+	d.sessionRequests <- true
```
Some goroutines block here
```
48 @ 0x4031fef 0x4007543 0x4007519 0x4007295 0x4742a0e 0x4742884 0x4961a3e 0x495d15d 0x4961c7c 0x495be6c 0x4946d9c 0x49467e0 0x4945adf 0x495f437 0x44812da 0x44876a1 0x4497e31 0x4061081
#	0x4742a0d	github.com/getlantern/lampshade.(*dialer).getSession+0x7d		/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190910214900-cfa0ddb2100f/dialer.go:184
#	0x4742883	github.com/getlantern/lampshade.(*dialer).DialContext+0x53		/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190910214900-cfa0ddb2100f/dialer.go:173
...
#	0x4497e30	net/http.(*Transport).getConn.func4+0xd0				/usr/local/Cellar/go/1.12.6/libexec/src/net/http/transport.go:1015
...

10 @ 0x4031fef 0x4007543 0x4007519 0x4007295 0x4742a0e 0x4742884 0x4961a3e 0x495d15d 0x4961c7c 0x495be6c 0x4946d9c 0x49467e0 0x4945adf 0x471ae03 0x471a4a5 0x4719c44 0x4a91b43 0x4a92632 0x4a8c736 0x4a8be97 0x4758152 0x475ad46 0x4759743 0x475878f 0x4a8e79d 0x4061081
#	0x4742a0d	github.com/getlantern/lampshade.(*dialer).getSession+0x7d			/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190910214900-cfa0ddb2100f/dialer.go:184
#	0x4742883	github.com/getlantern/lampshade.(*dialer).DialContext+0x53			/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190910214900-cfa0ddb2100f/dialer.go:173
...
#	0x4a8e79c	github.com/getlantern/flashlight/client.(*Client).handle+0x19c			/Users/joesis/workspace/go/src/github.com/getlantern/flashlight/client/handler.go:33
```

joesis replied:
Okay so if `s.getSession()` times out, `recycleSession` won't be called. If it happened for 10 times, the `sessionRequests` would be full, cause all later `s.getSession()` timing out as well. This patch fixed it for me, though the role of `sessionRequests` is not very obvious at first sight.

```
diff --git a/dialer.go b/dialer.go
index 1004025..f4621e2 100644
--- a/dialer.go
+++ b/dialer.go
@@ -171,13 +171,11 @@ func (d *dialer) Dial() (net.Conn, error) {

 func (d *dialer) DialContext(ctx context.Context) (net.Conn, error) {
        s, err := d.getSession(ctx)
+       go d.recycleSession(s)
        if err != nil {
                return nil, err
        }
-       c := s.createStream(ctx)
-       go d.recycleSession(s)
-
-       return c, nil
+       return s.createStream(ctx), nil
 }
```

myleshorton replied:
OK fixed! Another great catch! There might be some simpler way, but `sessionRequests` is just intended to trigger `recycleSessions` into knowing there's an active request, and that it's therefore safe to put the current session in `liveSessions`. That way `liveSessions` never has almost never contains closed sessions, and the only time it can have a closed session is if a session is closed between `recycleSessions` putting it in `liveSessions` and `getSession` pulling it out of `liveSessions`, but that will be caught by the `isClosed` check.

joesis replied:
Ha I realized I introduced another bug which could call `recycleSession` with an invalid session, but luckily you fixed it.

oxtoacart commented on dialer.go:
```diff
@@ -68,175 +81,177 @@ func NewDialer(opts *DialerOpts) Dialer {
 
-func (bd *boundDialer) DialContext(ctx context.Context) (net.Conn, error) {
-	return bd.Dialer.DialContext(ctx, bd.dial)
+	if err != nil {
```
Should close conn if unable to start session. Again, not super likely but better to be clean about it.

myleshorton replied:
Yup good catch fixed!

oxtoacart commented on dialer.go:
```diff
@@ -68,175 +81,177 @@ func NewDialer(opts *DialerOpts) Dialer {
+	// disruptions.
+	d.pendingSessions <- &sessionConfig{
+		name:         "liveness to " + d.name,
+		dialTimeout:  5 * time.Second,
```
Would be good to make this configurable. Also, since we're doing these in the background anyway, I think a bigger timeout might be good for high congestion periods.

oxtoacart replied:
Oh never mind, I see that the main one has a much longer dial timeout. Nice!

myleshorton commented:
@joesis I made a tweak to eliminate the `sessionRequests` channel and to just use blocking channels to simplify this slightly. What do you think?

joesis commented on dialer.go:
```diff
@@ -68,175 +88,178 @@ func NewDialer(opts *DialerOpts) Dialer {
+	// Selecting on those two cases avoids closed sessions building up in the live session channel.
+	select {
+	case <-s.getCloseCh():
+		d.log.Debugf("Session closed before requested. Requesting new session: %#v", s.getSessionConfig())
```
I got around 10 such messages after running for 4 minutes, suggested that some sessions were wasted. Checking in detail...

joesis replied:
Those were idle closed. Not new, but better to avoid creating sessions if the dialer is never selected.

The current code avoided this by starting session on demand, with the cost of taking longer to switch to an idled proxy, because dialing a new session is required. OTOH, when more than one proxy look fail / get blocked at the same time, a pre-established session is nothing but an illusion.

joesis replied:
45 "Session closed before requested" when running for 30 minutes, of which 40 (30m / 1m30s * 2) were from a proxy never selected. Not that bad.

joesis commented on dialer.go:
```diff
@@ -68,175 +88,178 @@ func NewDialer(opts *DialerOpts) Dialer {
-			newSession(minLiveConns)
-		case <-time.After(d.redialSessionInterval):
-			newSession(d.maxLiveConns)
+			d.log.Debugf("Returning session in %v", time.Since(start))
```
Logging for each dial attempt seems excessive. Maybe change to `Tracef`?

myleshorton replied:
Yeah sorry that wasn't meant to make it into the final version - removing!

joesis commented:
Looks good and is running good @myleshorton , with the only concern it will waste some resources keeping never used sessions.

myleshorton commented:
Thanks @joesis. The extra connections are interesting. I don't think it's a huge deal, although it will increase load on proxies (although particularly on proxies that are likely to be underutilized). To me this is potentially advantageous when things start to break down, as Lantern ideally is able to quickly switch to a connection with less congestion (if that exists -- depends where the congestion is on the line I suppose).

joesis commented:
Yeah totally make sense!

joesis commented:
Last time I saw it, my Lantern had around 800 goroutines of which mainly were lampshade ones. I thought it was acceptable as they were scattered in different code paths, but I was wrong. After my Mac slept for several hours, Lantern was killed for exceeding 8128 goroutines. Will watch to see why the number is higher than expected.

myleshorton commented:
Interesting thanks @joesis I'll have a look as well!
